src/metabase/models/card.clj

```diff
@@ -344,8 +361,13 @@
       ;; make sure this Card doesn't have circular source query references if we're updating the query
       (when (:dataset_query changes)
         (check-for-circular-source-query-references changes))
+      ;; updating a model dataset query to includes any clauses will disable implicit acitons(if it exists)
+      (when (and (:dataset_query changes)
+                 (:dataset old-card-info)
+                 (not (model-supports-implicit-actions? changes)))
+        (disable-implicit-action-for-model! id))
       ;; Archive associated actions
-      (when (and (not (:dataset changes))
+      (when (and (false? (:dataset changes))
```

Comment:
this was a bug before and tested it here : https://github.com/metabase/metabase/blob/fe252f7c79a8584b18a10a195d9007acd8983a6b/test/metabase/models/action_test.clj#L137